AEPMessaging/Sources/MessagingRulesEngine.swift

```diff
@@ -44,45 +45,53 @@ class MessagingRulesEngine {
-                Log.debug(label: MessagingConstants.LOG_TAG, "Skipping proposition with malformed in-app message content.")
-                continue
+        var tempPropInfo: [String: PropositionInfo] = [:]
+        
+        if let propositions = propositions {
+            for proposition in propositions {
+                guard let ruleString = proposition.items.first?.data.content else {
+                    Log.debug(label: MessagingConstants.LOG_TAG, "Skipping proposition with no in-app message content.")
+                    continue
+                    
+                }
+                
+                guard let rule = processRule(ruleString) else {
+                    Log.debug(label: MessagingConstants.LOG_TAG, "Skipping proposition with malformed in-app message content.")
+                    continue
+                }
+                
+                // pre-fetch the assets for this message if there are any defined
+                cacheRemoteAssetsFor(rule)
+                
+                // store reporting data for this payload for later use
+                if let messageId = rule.first?.consequences.first?.id {
+                    tempPropInfo[messageId] = proposition.propositionInfo
+                }
+                
+                rules.append(contentsOf: rule)
             }
-
-            // pre-fetch the assets for this message if there are any defined
-            cacheRemoteAssetsFor(rule)
-            // store reporting data for this payload for later use
-            storePropositionInfo(proposition, forMessageId: rule.first?.consequences.first?.id)
-
-            rules.append(contentsOf: rule)
         }
 
-        rulesEngine.replaceRules(with: rules)
-        Log.debug(label: MessagingConstants.LOG_TAG, "Successfully loaded \(rules.count) message(s) into the rules engine.")
-    }
-    
-    func clearPropositions() {
-        propositionInfo.removeAll()
-        clearPropositionsCache()
-        rulesEngine.replaceRules(with: [])        
-        Log.debug(label: MessagingConstants.LOG_TAG, "In-app messages cleared from Messaging rules engine and NSUserDefaults.")
-    }
-
-    func storePropositionInfo(_ proposition: PropositionPayload, forMessageId messageId: String?) {
-        guard let messageId = messageId else {
-            Log.debug(label: MessagingConstants.LOG_TAG, "Unable to associate proposition information for in-app message. MessageId unavailable in rule consequence.")
-            return
+        if clearExisting {
+            inMemoryPropositions.removeAll()
+            cachePropositions(nil)
+            propositionInfo = tempPropInfo
+            rulesEngine.replaceRules(with: rules)
+            Log.debug(label: MessagingConstants.LOG_TAG, "Successfully loaded \(rules.count) message(s) into the rules engine.")
+        } else {
+            propositionInfo.merge(tempPropInfo) { _, new in new }
+            rulesEngine.addRules(rules)
```

Comment:
Is this a no-op if invoked with empty rules ([]) or we need to make a prior check?

Reply:
if we aren't clearing existing and there are no new rules, it is essentially a no-op